src/flp/types/dp.rs

```diff
@@ -0,0 +1,327 @@
+use crate::dp::{distributions::PureDpDiscreteLaplace, DifferentialPrivacyStrategy};
+use crate::dp::{DifferentialPrivacyDistribution, DpError};
+use crate::field::{FftFriendlyFieldElement, Field128, Field64};
+use crate::flp::gadgets::{Mul, ParallelSumGadget};
+use crate::flp::types::{Histogram, SumVec};
+use crate::flp::{FlpError, TypeWithNoise};
+use crate::vdaf::xof::SeedStreamTurboShake128;
+use num_bigint::{BigInt, BigUint, TryFromBigIntError};
+use num_integer::Integer;
+use num_traits::{One, Pow};
+use rand::{distributions::Distribution, Rng, SeedableRng};
+
+// TODO(#1071): This is implemented for the concrete fields `Field64` and `Field128` in order to
+// avoid imposing the `BigInt: From<F::Integer>` bound on all callers. In the future, we may want to
+// impose this bound on `FieldElementWithInteger`, which would allow simple blanket implementations
+// here.
+impl<S> TypeWithNoise<PureDpDiscreteLaplace> for SumVec<Field64, S>
+where
+    S: ParallelSumGadget<Field64, Mul<Field64>> + Eq + 'static,
+{
+    fn add_noise_to_result(
+        &self,
+        dp_strategy: &PureDpDiscreteLaplace,
+        agg_result: &mut [Self::Field],
+        _num_measurements: usize,
+    ) -> Result<(), FlpError> {
+        self.add_noise(
+            dp_strategy,
+            agg_result,
+            &mut SeedStreamTurboShake128::from_entropy(),
+        )
+    }
+}
+
+impl<S> TypeWithNoise<PureDpDiscreteLaplace> for SumVec<Field128, S>
+where
+    S: ParallelSumGadget<Field128, Mul<Field128>> + Eq + 'static,
+{
+    fn add_noise_to_result(
+        &self,
+        dp_strategy: &PureDpDiscreteLaplace,
+        agg_result: &mut [Self::Field],
+        _num_measurements: usize,
+    ) -> Result<(), FlpError> {
+        self.add_noise(
+            dp_strategy,
+            agg_result,
+            &mut SeedStreamTurboShake128::from_entropy(),
+        )
+    }
+}
+
+impl<F, S> SumVec<F, S>
+where
+    F: FftFriendlyFieldElement,
+    BigInt: From<F::Integer>,
+    F::Integer: TryFrom<BigInt, Error = TryFromBigIntError<BigInt>>,
+{
+    fn add_noise<R>(
+        &self,
+        dp_strategy: &PureDpDiscreteLaplace,
+        agg_result: &mut [F],
+        rng: &mut R,
+    ) -> Result<(), FlpError>
+    where
+        R: Rng,
+    {
+        // Compute the l1-sensitivity of the aggregation function (assuming the substitution-DP
+        // model). The worst case is when one individual's measurement changes such that each vector
```

Comment:
I'm not sure that substitution-DP is relevant to this case.  Your goal is to measure the contribution of a single submission.  A submission could be $[0..2^b)$ on all dimensions equally, which leads to this outcome.  But that does not imply substitution.  If you have substitution, then you could have $\\{A=0 , B=2^b-1\\}$ or $\\{A=2^b-1 , B=0\\}$.  That means sensitivity under substitution effectively doubles, but I don't see how substitution applies to DAP.

Reply:
I relied on the reasoning in https://wangshan.github.io/draft-wang-ppm-differential-privacy/draft-wang-ppm-differential-privacy.html#appendix-B.2 for the choice of substitution-DP. The collector also learns the number of reports (which is needed for unsharding in some VDAFs, or for RAPPOR-style debiasing) and thus we can only get indistinguishable protocol runs if we assume substitution.

Reply:
I don't agree with that reasoning.  It will depend on the setting, but you have to assume that each contribution is unknown to the collector.  A histogram contribution could be a 0 just as much as it could be a 1.  That's the contribution that we're hiding, not that a given person is participating or not.  Substitution assumes that each report contributes something, which is not something we rely on.

My suggestion would be to externalize whether you want to use substitution as the basis of your sensitivity analysis.  That is, if someone believes that substitution is necessary in their setting, they can set epsilon accordingly (i.e., halve it).  Otherwise, settings where it is directly detrimental (i.e., the case we currently care about) will have to adjust in the other direction.

Reply:
In the case of `Prio3Histogram`, the validity circuit enforces that every contribution adds 1 to one of the buckets, so there is no possible 'zero measurement' to exchange with. Using `Prio3MultihotCountVec` with `max_weight = 1` instead would permit contributions to add nothing.

We could allow a choice between substitution and "replacement with zero" (as a modification of deletion-DP) by rolling that choice into the `DifferentialPrivacyStrategy` trait as well. Then, the different strategy-specific implementations of `AggregatorWithNoise`/`TypeWithNoise` could use their own sensitivity formulas as appropriate.

Reply:
Agree with this reasoning for the choice of substitution-DP:

> The collector also learns the number of reports (which is needed for unsharding in some VDAFs, or for RAPPOR-style debiasing) and thus we can only get indistinguishable protocol runs if we assume substitution.

For unsharding in VDAF, `num_measurements` is an argument in the public API, and in DAP, `report_count` is in [`Collection`](https://www.ietf.org/archive/id/draft-ietf-ppm-dap-11.html#section-4.6.1-17), so the number of reports in the batch is public information.

> My suggestion would be to externalize whether you want to use substitution as the basis of your sensitivity analysis. That is, if someone believes that substitution is necessary in their setting, they can set epsilon accordingly (i.e., halve it).

If the code assumes substitution-DP in its sensitivity analysis, and if someone wants to use deletion-DP, they can double the target central epsilon?

Reply:
Does @martinthomson's suggestion imply specific client behavior, such as randomized response? For example, a client might submit a 1-hot vector, or a vector of all 0s, depending on the outcome of a coinflip?

Exposing in the API what notion of "neighboring databases" we consider seems fraught to me. It's perhaps safer if we don't ask the user to go through this analysis themselves. Of course, this is just applying a more general principle of good API design: with DP we may be in a situation where we really do need to expose a choice (and have an expert make that choice).

Question: Is there a way to be "conservative" here? In particular, is it the case that substitution-DP implies deletion-DP with the same bound, but perhaps at the cost of some amount of utility?

I understand there is time pressure to merge this, so if we merge as-is, I would suggest at least making sure it's documented that we're assuming the application requires substitution-DP.

Reply:
The neighboring dataset definitions used are documented on the `PureDpDiscreteLaplace` and `ZCdpDiscreteGaussian` items now.

The comparison between substitution versus deletion differs based on the query -- in this PR, you get the same numeric global sensitivity value either way for Prio3SumVec, whereas you get 2 or 1 for Prio3Histogram. I think it will always be the case that you get sensitivities that are greater or equal with substitution, since a substitution is equivalent to a deletion and an insertion. However, the different assumptions complicate things, because an epsilon value used in a substitution-DP mechanism measures against a different set of counterfactuals than an epsilon value used in a deletion-DP mechanism.